lib/request/complement.js

```diff
@@ -100,34 +64,52 @@ module.exports = function complement(requested, optimized, depthDifferences, tre
- *  - `depth`: index of optimized path that we are trying to match with `currentTree`
- *  - `rCurrentPath`: current accumulated requested path by previous recursive
- *                    iterations. Could also have been pre-initialized as stated
- *                    above.
- *                    This path cannot contain ranges, instead contains a key
- *                    from the range, representing one of the individual paths
- *                    in `requestedPath` pathset
- *  - `oCurrentPath`: corresponding accumulated optimized path, to be matched
- *                    with `currentTree`. Could have been pre-initialized.
- *                    Cannot contain ranges, instead contains a key from the
- *                    range at given `depth` in `optimizedPath`
- *  - `depthDiff`: difference in length between `requestedPath` and `optimizedPath`
  *
- *  Example scenario:
+ * Example scenario:
  *      - requestedPath: ['lolomo', 0, 0, 'tags', { from: 0, to: 2 }]
  *      - optimizedPath: ['videosById', 11, 'tags', { from: 0, to: 2 }]
  *      - currentTree: { videosById: 11: { tags: { 0: null, 1: null }}}
- *      // since requested path is longer, optimized path index starts from depth 0
- *      // and accumulated requested path starts pre-initialized (rCurrentPath)
- *      - depth: 0
- *      - rCurrentPath: ['lolomo']
- *      - oCurrentPath: []
- *      - depthDiff: 1
+ *
+ * Returns a 3-tuple consisting of
+ *  - the intersection of requested paths with currentTree
+ *  - the complement of optimized paths with currenTree
+ *  - the complement of corresponding requested paths with currentTree
  */
-function findPartialIntersections(requestedPath, optimizedPath, currentTree, depth, rCurrentPath, oCurrentPath, depthDiff) {
+function findPartialIntersections(requestedPath, optimizedPath, currentTree) {
+    var depthDiff = requestedPath.length - optimizedPath.length;
+
+    // Pre-descend into the sub tree (when the optimized path is longer than the requested path)
+    for (var i = 0; currentTree && i < -depthDiff; i++) {
+        currentTree = currentTree[optimizedPath[i]];
+    }
+
+    // No matching path, cannot dedupe
+    if (!currentTree) {
+        return [[], [optimizedPath], [requestedPath]];
+    }
+
+    if (depthDiff === 0) {
+        return findPartialIntersectionsRec(requestedPath, optimizedPath, currentTree, 0, [], [], 0);
+    } else if (depthDiff > 0) {
+        return findPartialIntersectionsRec(requestedPath, optimizedPath, currentTree, 0, arraySlice(requestedPath, 0, depthDiff), [], depthDiff);
+    } else { // depthDiff < 0
+        return findPartialIntersectionsRec(requestedPath, optimizedPath, currentTree, -depthDiff, [], arraySlice(optimizedPath, 0, -depthDiff), depthDiff);
+    }
+}
+
+function findPartialIntersectionsRec(
+    requestedPath,
+    optimizedPath,
+    currentTree,
+    depth,
+    rCurrentPath,
+    oCurrentPath,
+    depthDiff
```

Comment:
`findPartialIntersectionsRec` has an inner `depthDiff` as well as an outer one. Can feel confusing to someone reading the code. If you're not modifying its value, might be worth just keeping `depthDiff` in the outer scope alone.

Reply:
my bad, should've been reading the code as a file. missed indentation in the diff view, and thought `findPartialIntersectionsRec` is defined _inside_ `findPartialIntersection`